driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -150,13 +170,73 @@ private void buildTableMetadata(KeyspaceMetadata ksm, List<Row> cfRows, Map<Stri
                         cols = Collections.<String, ColumnMetadata.Raw>emptyMap();
                     }
                 }
-                TableMetadata.build(ksm, cfRow, cols, cassandraVersion);
+
+                TableMetadata tm = TableMetadata.build(ksm, cfRow, cols, cassandraVersion);
+                TableMetadata tmPrevious = ksm.tables.get(tm.getName());
+                if (tmPrevious == null) {
+                    ksm.add(tm);
+                    triggerOnTableAdded(tm);
```

Comment:
I agree to leave it the way it is. If listeners block for an unreasonable amount of time, it's going to block _some_ driver thread anyway. I'll amend the doc to mention that listeners should not block.